sql: add cluster version for udfs with mutations

rharding6373 opened the pull request:
sql: move udf mutations privileges logic tests to their own file
    
Epic: None
    
Release note: None

sql: add cluster version for udfs with mutations

This PR gates the UDFs containing mutations feature.

Epic: CRDB-25388
Informs: #105365

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112148)
<!-- Reviewable:end -->

rharding6373 committed:
sql: move udf mutations privileges logic tests to their own file

Epic: None

Release note: None

rharding6373 committed:
sql: add cluster version for udfs with mutations

This PR gates the UDFs containing mutations feature.

Epic: CRDB-25388
Informs: #105365

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 20 of 20 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112148#-:-NgVddk_B3GnZwXwMd3_:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/builder.go` line 331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112148#-NgVdKba61IUaECgJamU:-NgVdKba61IUaECgJamV:bnn5xbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c129c156317ca652505ca9edc3142200d58f6cd2/pkg/sql/opt/optbuilder/builder.go#L331)):*
> ```Go
> 		case *tree.Insert, *tree.Update, *tree.Delete:
> 			activeVersion := b.evalCtx.Settings.Version.ActiveVersion(b.ctx)
> 			if !activeVersion.IsActive(clusterversion.V23_2_PLpgSQL) {
> ```

`V23_2_PLpgSQL` -> `V23_2_UDFMutations`


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
TFTR! Only test failure is a `pq: validate fk constraint: unsupported comparison operator: <refcursor> = <refcursor>` error on a randomized test, which is unrelated and already has an issue open.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112148#-:-NgWuISRElYY9vV3Xv7-:by92ttz)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/optbuilder/builder.go` line 331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112148#-NgVdKba61IUaECgJamU:-NgWuDAd4F_7XscMbIH8:b-hcxwq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c129c156317ca652505ca9edc3142200d58f6cd2/pkg/sql/opt/optbuilder/builder.go#L331)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

`V23_2_PLpgSQL` -> `V23_2_UDFMutations`
</blockquote></details>

Ouch, glad you caught that.


<!-- Sent from Reviewable.io -->
